controllers/clustertopology_controller_compute_test.go

```diff
@@ -250,7 +252,281 @@ func TestGetClass(t *testing.T) {
+				infrastructureCluster: nil,
+				machineDeployments:    emptyMachineDeployments,
+			},
+		},
+		{
+			name:    "Cluster with non existent infrastructure reference only",
+			cluster: newFakeCluster(metav1.NamespaceDefault, "cluster1").withInfraRef(objToRef(nonExistentInfraCluster)).Obj(),
+			objects: []client.Object{
+				infraCluster,
+			},
+			wantErr: true, // this test fails as partial reconcile is undefined.
+		},
+		{
+			name:    "Cluster infrastructure reference only",
+			cluster: newFakeCluster(metav1.NamespaceDefault, "cluster1").withInfraRef(objToRef(infraCluster)).Obj(),
+			objects: []client.Object{
+				infraCluster,
+			},
+			// Expecting valid return with no control plane or machine deployment state defined but with a valid infrastructure state
+			want: &clusterTopologyState{
+				cluster:               newFakeCluster(metav1.NamespaceDefault, "cluster1").withInfraRef(objToRef(infraCluster)).Obj(),
+				controlPlane:          controlPlaneTopologyState{},
+				infrastructureCluster: infraCluster,
+				machineDeployments:    emptyMachineDeployments,
+			},
+		},
+		{
+			name:    "Cluster with infrastructure reference and control plane reference, no control plane infrastructure and a clusterClass with no infra requirement",
+			cluster: newFakeCluster(metav1.NamespaceDefault, "cluster1").withControlPlaneRef(objToRef(controlPlane)).withInfraRef(objToRef(infraCluster)).Obj(),
+			class:   *clusterClassWithNoControlPlaneInfra,
+			objects: []client.Object{
+				controlPlane,
+				infraCluster,
+				clusterClassWithNoControlPlaneInfra,
+			},
+			// Expecting valid return with control plane with no control plane infrastructure state, cluster infrastructure state and no defined machine deployment state
+			want: &clusterTopologyState{
+				cluster:               newFakeCluster(metav1.NamespaceDefault, "cluster1").withControlPlaneRef(objToRef(controlPlane)).withInfraRef(objToRef(infraCluster)).Obj(),
+				controlPlane:          controlPlaneTopologyState{object: controlPlane, infrastructureMachineTemplate: nil},
+				infrastructureCluster: infraCluster,
+				machineDeployments:    emptyMachineDeployments,
+			},
+		},
+		{
+			name:    "Cluster with infrastructure reference and control plane reference, no control plane infrastructure and a clusterClass with an infra requirement",
+			cluster: newFakeCluster(metav1.NamespaceDefault, "cluster1").withControlPlaneRef(objToRef(controlPlane)).withInfraRef(objToRef(infraCluster)).Obj(),
+			class:   *clusterClassWithControlPlaneInfra,
+			objects: []client.Object{
+				controlPlane,
+				infraCluster,
+				clusterClassWithControlPlaneInfra,
+			},
+			// Expecting error from control plane having no valid control plane infrastructure with clusterclass requiring control plane infrastructure.
+			wantErr: true,
+		},
+		{
+			name:    "Cluster with control plane reference and with control plane infrastructure, but no cluster infrastructure",
```

Comment:
I'm probably missing something, but why doesn't this lead to a panic when class is not set?

Reply:
Great catch! this is actually a panic when I try to access it directly during the test. I'm not sure why that doesn't happen during the run of the full function. This was an oversight on my part for sure - the class is assumed to exist so we don't actually do a nil check for it or anything.

Reply:
As far as I can see (and this took me too long!) It's because we dereference this pointer before passing the class to the the function. That means we end up with an empty struct of type v1.ClusterClass which has the right fields, they're just set to nil. That's a valid value for the field in the code, so we continue through with no error. I will add a clusterclass to this test case though.

Reply:
Ups, yeah we take a struct. I somehow just assumed it's a struct pointer :)

Reply:
Not sure if it maybe should be a pointer. But I don't have a strong opinion/arguments either way :)